sklearn/metrics/tests/test_ranking.py

```diff
@@ -693,6 +696,107 @@ def test_score_scale_invariance():
+    ([0, 1, 1], [0, 0.25, 0.5], [0], [0]),
+    ([0, 1, 1], [0.5, 0.75, 1], [0], [0]),
+    ([0, 1, 1], [0.25, 0.5, 0.75], [0], [0]),
+    ([1, 0, 0], [0, 0.5, 1], [1, 1, 0.5], [0, 1, 1]),
+    ([1, 0, 0], [0, 0.25, 0.5], [1, 1, 0.5], [0, 1, 1]),
+    ([1, 0, 0], [0.5, 0.75, 1], [1, 1, 0.5], [0, 1, 1]),
+    ([1, 0, 0], [0.25, 0.5, 0.75], [1, 1, 0.5], [0, 1, 1]),
+    ([1, 0, 1], [0, 0.5, 1], [1, 1, 0], [0, 0.5, 0.5]),
+    ([1, 0, 1], [0, 0.25, 0.5], [1, 1, 0], [0, 0.5, 0.5]),
+    ([1, 0, 1], [0.5, 0.75, 1], [1, 1, 0], [0, 0.5, 0.5]),
+    ([1, 0, 1], [0.25, 0.5, 0.75], [1, 1, 0], [0, 0.5, 0.5]),
+])
+def test_detection_error_tradeoff_curve_toydata(
+    y_true, y_score, expected_fpr, expected_fnr
+):
+    # Check on a batch of small examples.
+    fpr, fnr, _ = detection_error_tradeoff_curve(y_true, y_score)
+
+    assert_array_almost_equal(fpr, expected_fpr)
+    assert_array_almost_equal(fnr, expected_fnr)
+
+
+@pytest.mark.parametrize("y_true,y_score,expected_fpr,expected_fnr", [
+    ([1, 0], [0.5, 0.5], [1], [0]),
+    ([0, 1], [0.5, 0.5], [1], [0]),
+    ([0, 0, 1], [0.25, 0.5, 0.5], [0.5], [0]),
+    ([0, 1, 0], [0.25, 0.5, 0.5], [0.5], [0]),
+    ([0, 1, 1], [0.25, 0.5, 0.5], [0], [0]),
+    ([1, 0, 0], [0.25, 0.5, 0.5], [1], [0]),
+    ([1, 0, 1], [0.25, 0.5, 0.5], [1], [0]),
+    ([1, 1, 0], [0.25, 0.5, 0.5], [1], [0]),
+])
+def test_detection_error_tradeoff_curve_tie_handling(
+    y_true, y_score, expected_fpr, expected_fnr
+):
+    fpr, fnr, _ = detection_error_tradeoff_curve(y_true, y_score)
+
+    assert_array_almost_equal(fpr, expected_fpr)
+    assert_array_almost_equal(fnr, expected_fnr)
+
+
+def test_detection_error_tradeoff_curve_sanity_check():
+    # Exactly duplicated inputs yield the same result.
+    assert_array_almost_equal(
+        detection_error_tradeoff_curve([0, 0, 1], [0, 0.5, 1]),
+        detection_error_tradeoff_curve(
+            [0, 0, 0, 0, 1, 1], [0, 0, 0.5, 0.5, 1, 1])
+    )
+
+
+@pytest.mark.parametrize("y_score", [
+    (0), (0.25), (0.5), (0.75), (1)
+])
+def test_detection_error_tradeoff_curve_constant_scores(y_score):
+    # Check computation with perfect and constant scores.
+    fpr, fnr, threshold = detection_error_tradeoff_curve(
+        y_true=[0, 1, 0, 1, 0, 1],
+        y_score=y_score * np.ones(6)
+    )
+
+    assert_array_almost_equal(fpr, [1])
+    assert_array_almost_equal(fnr, [0])
+    assert_array_almost_equal(threshold, [y_score])
+
+
+def test_detection_error_tradeoff_curve_bad_input():
+    # input variables with inconsistent numbers of samples
+    assert_raises(
+        ValueError,
+        detection_error_tradeoff_curve,
+        [0, 1], [0, 0.5, 1]
+    )
+    assert_raises(
+        ValueError,
+        detection_error_tradeoff_curve,
+        [0, 1, 1], [0, 0.5]
+    )
+
+    # When the y_true values are all the same a detection error tradeoff cannot
+    # be computed.
+    with np.errstate(all="raise"):
+        assert_raises(
+            FloatingPointError,
```

Comment:
would we be better off with a more explicit error message?

Reply:
I was considering it, but decided against it. 

Currently the function behaves like that: When you call detection_error_tradeoff_curve with all `y_true` the same it will return `NaN` in one of the return values and NumPy will throw a Warning. I find this behavior sort-of intuitive, as it suggests that the issue might be in the data and not in the way the function is called.

But I have no strong feeling towards either solution. I can change it, if this is standard behavior in other scikit-learn functions.

Reply:
My concern is that cases like this do happen, at least when the calling of the metric is automated within some cross-validation or similar routine (and hence data sub-samples are taken). If it is going to be hard for users to debug, we should give them a reasonably informative error message... if it doesn't involve excessive additional computation to handle their edge case.

Reply:
Good point. Given the computational overhead is indeed very small I will make this a little more explicit.